cli/internal/fs/package_deps_hash.go

```diff
@@ -31,354 +21,241 @@ type PackageDepsOptions struct {
+			return nil, err
 		}
-		key := filepath.ToSlash(relativePath)
-		result[key] = hash
+		relativeResult[UnsafeToRelativeUnixPath(relativePath)] = hash
 	}
-	return result, nil
-}
 
-// threadsafeBufferWriter is a wrapper around a byte buffer and a lock, allowing
-// multiple goroutines to write to the same buffer. No attempt is made to
-// lock around reading, which should only be done once no more writing will occur.
-type threadsafeBufferWriter struct {
-	mu     sync.Mutex
-	buffer bytes.Buffer
+	return relativeResult, nil
 }
 
-func (tsbw *threadsafeBufferWriter) Write(p []byte) (int, error) {
-	tsbw.mu.Lock()
-	defer tsbw.mu.Unlock()
-	return tsbw.buffer.Write(p)
-}
+// gitHashObject takes a list of files returns a map of with their git hash values.
+// It uses git hash-object under the hood.
+// Note that filesToHash must have full paths.
+func gitHashObject(rootPath AbsolutePath, filesToHash []string) (map[RelativeUnixPath]string, error) {
+	fileCount := len(filesToHash)
+	changes := make(map[RelativeUnixPath]string, fileCount)
 
-var _ io.Writer = (*threadsafeBufferWriter)(nil)
+	if fileCount > 0 {
+		cmd := exec.Command("git", "hash-object", "--stdin-paths")
+		cmd.Dir = rootPath.ToString()
 
-// gitHashForFiles a list of files returns a map of with their git hash values. It uses
-// git hash-object under the hood.
-// Note that filesToHash must have full paths.
-func gitHashForFiles(filesToHash []string) (map[string]string, error) {
-	changes := make(map[string]string)
-	if len(filesToHash) > 0 {
-		input := []string{"hash-object"}
-		input = append(input, filesToHash...)
-		cmd := exec.Command("git", input...)
-		// exec.Command writes to stdout and stderr from different goroutines,
-		// but only if they aren't the same io.Writer. Since we're passing different
-		// io.Writer instances that can optionally write to the same buffer, we need
-		// to do the locking ourselves.
-		var allout threadsafeBufferWriter
-		var stdout bytes.Buffer
-		// write stdout to both the stdout buffer, and the combined output buffer
-		// we don't expect to need stderr, but in the event that something goes wrong,
-		// we'd like to report the entirety of the output in the order it was output.
-		mw := io.MultiWriter(&allout, &stdout)
-		cmd.Stdout = mw
-		cmd.Stderr = &allout
-		err := cmd.Run()
-		if err != nil {
-			output := string(allout.buffer.Bytes())
-			return nil, fmt.Errorf("git hash-object exited with status: %w. Output:\n%v", err, output)
+		stdinPipe, stdinPipeError := cmd.StdinPipe()
+		if stdinPipeError != nil {
+			return nil, stdinPipeError
+		}
+
+		go func() {
+			defer func() {
+				stdinPipeCloseError := stdinPipe.Close()
+				if stdinPipeCloseError != nil {
+					return
+				}
+			}()
+
+			for _, file := range filesToHash {
+				// Expects paths to be one per line so we escape newlines
+				_, err := io.WriteString(stdinPipe, strings.ReplaceAll(file, "\n", "\\n")+"\n")
```

Comment:
> Can you add a comment as to why we're using stdin and piping in files, rather than passing them as commandline arguments?

Until I get them into the code:
1. This should fix the most common cause of #1097: argument size limitations.
2. Only newlines need to be escaped.
3. Streaming! It can start processing even while `stdin` isn't closed!

> Additionally, is there a reason this can't or shouldn't use runGitCommand?

I broke the pattern with this one because the structure of the output is quite distinct. It doesn't actually end up down in `gitoutput` and has its own usage of `scanner.Scan`.